src/v/cluster/archival/archiver_scheduler_impl.cc

```diff
@@ -0,0 +1,152 @@
+        v->backoff = v->backoff.value_or(0ms)
+                     + v->backoff.value_or(_initial_backoff());
+        v->backoff = std::clamp(v->backoff.value(), 0ms, _max_backoff());
+
+        auto sleep_time = v->backoff.value();
+        auto now = Clock::now();
+        if (
+          arg.next_housekeeping.has_value()
+          && arg.next_housekeeping.value() > now) {
+            // In case if the ntp_archiver reported its next
+            // housekeeping timestamp or next manifest reupload timestamp
+            // we can override the sleep_time to wake up early.
+            // This will not affect the backoff time (which will affect
+            // next sleep). It will only replace current sleep time.
+            auto dt = arg.next_housekeeping.value() - now;
+            if (dt < sleep_time) {
+                sleep_time
+                  = std::chrono::duration_cast<std::chrono::milliseconds>(dt);
+            }
+        }
+        if (
+          arg.next_manifest_upload.has_value()
+          && arg.next_manifest_upload.value() > now) {
+            auto dt = arg.next_manifest_upload.value() - now;
+            if (dt < sleep_time) {
+                sleep_time
+                  = std::chrono::duration_cast<std::chrono::milliseconds>(dt);
+            }
+        }
+
+        // Backoff is applied in case of any error including no-data error.
+        try {
+            co_await ss::sleep_abortable<Clock>(
+              sleep_time, arg.archiver_rtc.get().root_abort_source());
+        } catch (...) {
+            auto er = std::current_exception();
+            if (ssx::is_shutdown_exception(er)) {
+                co_return error_outcome::shutting_down;
+            }
+            vlog(archival_log.error, "Unexpected sleep error {}", er);
+        }
+    } else {
+        // No error so we can proceed to next upload immediately
+        // if rate limits are not reached.
+        v->backoff.reset();
+    }
+    auto puts = arg.put_requests_used > 0 ? arg.put_requests_used : 1;
+    try {
+        co_await _put_requests.throttle(
+          puts, arg.archiver_rtc.get().root_abort_source());
+        co_await _shard_tput_limit.throttle(
+          arg.uploaded_bytes, arg.archiver_rtc.get().root_abort_source());
+    } catch (...) {
+        auto er = std::current_exception();
+        if (ssx::is_shutdown_exception(er)) {
+            co_return error_outcome::shutting_down;
+        }
+        vlog(archival_log.error, "Unexpected throttling error {}", er);
+        co_return error_outcome::unexpected_failure;
+    }
+
+    upload_resource_quota res{
+      .requests_quota = _put_requests.available(),
+      .upload_size_quota = _shard_tput_limit.available(),
+    };
+    co_return res;
+}
+
+template<class Clock>
+ss::future<> archiver_scheduler<Clock>::create_ntp_state(model::ntp ntp) {
+    auto it = _partitions.find(ntp);
+    vassert(it == _partitions.end(), "Partition {} is already scheduled", ntp);
```

Comment:
What guarantees this? Can you add a comment in the header explaining to callers/readers what should be called before this to ensure this doesn't get hit? My guess: this is only expected to be called once before eventually calling dispose_ntp_state()?